묵찌빠 게임 [STEP 1] Minsup, Whales

agilestarskim opened the pull request:
# 묵찌빠 게임 [STEP 1] Minsup, Whales

### saleyMinsup조

안녕하세요 릴라(@juun97)!!
릴라🦍의 두번째 리뷰이가 된 Minsup🦊, Whales🐋입니다!!
이번주 묵찌빠 게임 리뷰 잘 부탁드립니다 🙏🏻

### 고민한 점
> 공부의 순서
- 공식 문서 vs 미션
처음에는 STEP1만 올라온 미션지를 보고 어렵지 않을 것 같다는 생각에 첫 날부터 미션을 하려고 계획했었습니다. 그러다 공식 문서를 꼼꼼히 봐야하고, 미션보다 공부가 중요하다는 윗분들 말씀에 미션 보다 공부를 먼저하고 그 공부를 같이 하는 쪽으로 방향을 잡았습니다!
    <details>
    <summary>결론</summary>
    정말 너---무 잘 했다고 생각합니다!! 어제 공부한 부분을 오늘 미션 코드에 유용하게 써먹을 수 있어서 효율적인 코드가 나와서 좋고, 공부하는 재미를 직접 느낄 수 있어서 신이 났습니다!!
    </details>

> 코드
- 게임종료 시점 고민
    - `while true` 반복문 안에 `switch`문으로 게임을 시작
    - `0` 입력시 게임종료이므로 반복문을 닫아야하는데 `case` 안에서 `break` 입력 시
`switch`문만 빠져나오고 반복문이 멈춰지지않는다.
    <details>
    <summary>해결</summary>
    
    어제 같이 공부한 `while loop 이름` 활용
    -> `while`에 이름을 붙이고 `break 이름`하면 `switch`문에서 `while`문 `break`가능
    ![](https://i.imgur.com/cxEh70H.png)
    ![](https://i.imgur.com/NSOajqq.png)
    </details>
- `가위(1)`가 `보(3)`를 이기는 조건
    - 처음에는 그냥 크기 비교로 조건을 짰다가 `가위vs보`에서 잘못된 결과가 나오는 것을 알고 고민
    <details>
    <summary>코드</summary>

    ```swift
    func check(input: String?, computer: String) {
        if let input2 = input, input2 == computer {
            print("비겼습니다!")
        } else if let input2 = input, input2 > computer {
            print("이겼습니다!")
        } else {
            print("졌습니다!")
        }
    }
    ```
    </details>

    - 상세한 `if` 구문으로 수정
    <details>
    <summary>코드</summary>

    ```swift
    func showResult(input: HandShape, computer: HandShape) {
    if input == .rock {
        if computer == .rock {
            print("비겼습니다!")
        } else if computer == .scissors {
            print("이겼습니다!")
        } else {
            print("졌습니다!")
        }
    } else if input == .scissors {
        …
    ```
    </details>

    - `튜플`로 수정
    <details>
    <summary>코드</summary>

    ```swift
    func isOver(with computer: String, by user: String) -> Bool {
        switch (user, computer) {
        case ("1", "1"), ("2", "2"), ("3", "3"):
            print("비겼습니다!")
            return false
        case ("1", "3"), ("3", "2"), ("2", "1"):
            print("이겼습니다!")
            return true
        default:
            print("졌습니다!")
            return true
        }
    }
    ```
    </details>


### 조언을 얻고 싶은 부분
> 코드
- 승패를 판정하는 함수의 이름 고민
    - 현재 `judgeGame(user, with: computer)`로 함수가 선언되어있는데, 함수의 이름과 내용이 완벽하게 일치하지 않는 것 같아 많은 고민을 했습니다.
    - 게임이 끝나는지 여부를 리턴하는 `isOver` 함수와 결과를 출력하는 `showResult`함수로 나누어볼까 생각했는데 오히려 가독성과 효율이 떨어지는 것 같아 `judgeGame`으로 병합했습니다. 
    - 릴라는 어떤 것이 더 나은 방법이라고 생각하시나요? 추가로 더 좋은 방법이 있으시면 조언 부탁드립니다.
- `rock scissor paper`를 `enum` 타입으로 만들지 고민
    - 처음에 `enum` 타입으로 만들어 사용했는데 막상 사용하는 부분이 적기도 하고 옵셔널 바인딩을 해야하는 과정이 복잡해서 `enum`을 사용하지 않고 숫자 그대로를 사용했습니다. 유지보수를 위해 `enum`으로 빼는 것이 나을까요?
    - 아니면 지금 당장 필요하지 않으니 숫자를 그대로 사용하는 것이 좋을까요?
    
```swift 
//예시 코드
enum HandShape: Int {
    case scissor = 1, rock, paper
}
let computer = HandShape(rawValue: String(Int.random(in: 1...3)!)!)!
let user = HandShape(rawValue: readLine()!)!
//옵셔널 바인딩을 두 세번 해야함 (예시코드의 간결성을 위해 강제 언래핑했음)
```

WhalesJin committed:
feat: 게임 실행 반복문 생성

Commit:
feat: 가위바위보 비교함수 생성

switch문 case 병합

WhalesJin committed:
fix: 가위바위보 비교함수 수정

함수명을 showResult로 바꾸고 보가 가위를 이기는 문제 해결

Commit:
refactor: 강제 언래핑에서 옵셔널 바인딩으로 변경

WhalesJin committed:
refactor: 가위바위보 비교 함수 수정

튜플을 이용한 비교 함수 수정

Commit:
refactor: 함수 이름 변경, 옵셔널 바인딩과 조건문 병합

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
 import Foundation
 
-print("Hello, World!")
+gameLoop: while true {
```
`labeled statement` 을 사용해주셨군요 👍🏻
저도 사용해보지 못한 문법인데 배워갑니다 🙇🏻‍♂️

juun97 commented:
안녕하세요 Minsup, Whales STEP 1 고생 많으셨습니다!
제가 드리는 의견들은 정답이 아닌 하나의 의견이라고 봐주세요 🙇🏻‍♂️


## 고민한 점
> 공부의 순서
>- 공식 문서 vs 미션
>처음에는 STEP1만 올라온 미션지를 보고 어렵지 않을 것 같다는 생각에 첫 날부터 미션을 하려고 계획했었습니다. 그러다 공식 문서를 꼼꼼히 봐야하고, 미션보다 공부가 중요하다는 윗분들 말씀에 미션 보다 공부를 먼저하고 그 공부를 같이 하는 쪽으로 방향을 잡았습니다!

- 되게 칭찬드리고 싶은 부분이에요!! 미션도 물론 중요하지만 미션의 과정에서 얻어가야 하는 부분을 확실히 내것으로 만드는게 더 중요하다고 생각합니다! 공식문서를 참고하는 습관 아주아주 최곱니다 💯

> 코드

- 코드에서 많이 고민하신 흔적이 남아있네요! 👍🏻 이런시도도 해보고 저런시도도 해보고 시행착오를 많이 겪어야 그 과정들에서 얻어가는게 많다고 생각합니다 😃
- 코드관련된 부분은 코멘트 드리면서 같이 얘기할게요!

## 조언을 얻고 싶은 부분

>- 승패를 판정하는 함수의 이름 고민
    - 현재 `judgeGame(user, with: computer)`로 함수가 선언되어있는데, 함수의 이름과 내용이 완벽하게 일치하지 않는 것 같아 많은 고민을 했습니다.
    - 게임이 끝나는지 여부를 리턴하는 `isOver` 함수와 결과를 출력하는 `showResult`함수로 나누어볼까 생각했는데 오히려 가독성과 효율이 떨어지는 것 같아 `judgeGame`으로 병합했습니다. 
    - 릴라는 어떤 것이 더 나은 방법이라고 생각하시나요? 추가로 더 좋은 방법이 있으시면 조언 부탁드립니다.

이 부분은 기능분리에 대해 더 고민을 해보시고 시도 해본다면 자연스럽게 해결이 되는 부분이라고 생각해요.
`하나의 메서드가 하나의 기능`만 할 수 있도록 한다면 이름은 그 기능에 맞춰 자연스럽게 지을 수 있을거에요. 😄 



>- `rock scissor paper`를 `enum` 타입으로 만들지 고민
    - 처음에 `enum` 타입으로 만들어 사용했는데 막상 사용하는 부분이 적기도 하고 옵셔널 바인딩을 해야하는 과정이 복잡해서 `enum`을 사용하지 않고 숫자 그대로를 사용했습니다. 유지보수를 위해 `enum`으로 빼는 것이 나을까요?
    - 아니면 지금 당장 필요하지 않으니 숫자를 그대로 사용하는 것이 좋을까요?

- 이 부분은 하단에서 얘기 드리겠습니다 😀


## 말씀드리고 싶은 부분

### 1️⃣ 기능분리

현재 main 파일에 모든 로직을 작성하셨는데, 그렇다면 main 파일의 기능은 무엇이라 생각하세요? 
현재 Xcode 의 ConsoleApp 에서 main 파일이 하는 역할에 대해 알아보시면 좋을것 같습니다.

위에서도 언급했다시피 하나의 메서드는 하나의 기능을 담당해야 한다고 많이들 말합니다. 

1. 그렇다면 왜 그래야 하는지 생각을 해본적이 있을까요? 

2. 왜 하나의 메서드가 여러기능을 담당하면 안될까요? 간단한 로직들이라면 하나의 메서드에서 다 처리 해버려도 별 문제가 없지 않을까요?

Minsup과 Whales의 생각을 한번 들려주시겠어요?


### 2️⃣ 매직넘버 / 리터럴

앞으로 많은 협업을 경험하게 될거에요. 그 과정에서 각자 `역할을 분리해 다른 부분을 담당`하게 될 수도 있고, `다른 사람의 작업물을 수정`하는 일을 할 수도 있어 내가 작성하지 않은 코드를 마주해야 할 일이 생기기 마련입니다.

현재 Minsup과 Whales이 작업하지 않았다고 가정하고 작업한 코드를 보았을 때 어떤 코드인지 알 수 있나요? 부가적인 설명없이 코드가 이해가 될까요?

짧은 코드가 당장에는 좋아보일 수도 있습니다. 하지만 짧아진 코드가 ` 자신만 알아볼 수 있는 코드`라면 과연 좋은 코드라고 할 수 있을까요?🤔



### 3️⃣ Branch

현재 main Branch 에서 작업을 진행해주셨네요! 혹시 branch 에 대해 알고계시나요? 

1. 하나의 branch 에서만 작업을 하게 된다면 어떤 문제가 생기게 될까요?

2. branch 를 나눠서 관리하면서 생기는 장점에 대해 알아볼까요

참고 - [브랜치란 무엇인가](https://git-scm.com/book/ko/v2/Git-%EB%B8%8C%EB%9E%9C%EC%B9%98-%EB%B8%8C%EB%9E%9C%EC%B9%98%EB%9E%80-%EB%AC%B4%EC%97%87%EC%9D%B8%EA%B0%80)

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+        print("게임 종료")
+        break gameLoop
+    case "1", "2", "3":
+        if let user = input, judgeGame(user, with: computer) {
+            break gameLoop
+        }
```
```suggestion
        if let user = input, 
           judgeGame(user, with: computer) {
            break gameLoop
        }
```
저 같은 경우 가독성을 위해 한줄의 하나의 로직만 위치하도록 작성을 하는 편입니다 😄

---

현재 judgeGame 의 결과가 true 여야 gameLoop 가 종료되고 있네요.
bool 타입으로 게임결과가 어떤 상태인지 명확하게 구분을 할 수 있을까요?

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+    }
+}
+
+func judgeGame(_ user: String, with computer: String) -> Bool {
```
Naming에 대한 질문사항이 있으셔서 조언드립니다.

`Naming `의 경우 어떤행동에 초점을 맞추는지에 따라 사람마다 작성하는 Naming 이 천차만별이라 생각합니다.
Minsup과 Whales 은 해당 메서드에 어떤 행동에 초점을 맞추고 judgeGame 이란 Naming 을 사용했을까요?

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+}
+
+func judgeGame(_ user: String, with computer: String) -> Bool {
+    switch (user, computer) {
```
Tuple 타입으로 비교 해주셨군요 👍🏻 가독성이 굉장히 좋은 것 같습니다 💯

juun97 replied:
Tuple 타입으로 비교 해주셨군요 👍🏻 가독성이 굉장히 좋은 것 같습니다 💯

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+
+func judgeGame(_ user: String, with computer: String) -> Bool {
+    switch (user, computer) {
+    case ("1", "1"), ("2", "2"), ("3", "3"):
```
언급했다 싶이 1, 2, 3 만 보고 해당 숫자가 어떤걸 의미하는지 알 수 있을까요?

매직넘버 / 리터럴에 대해 한번 알아보시겠어요?

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+func judgeGame(_ user: String, with computer: String) -> Bool {
+    switch (user, computer) {
+    case ("1", "1"), ("2", "2"), ("3", "3"):
+        print("비겼습니다!")
+        return false
+    case ("1", "3"), ("3", "2"), ("2", "1"):
+        print("이겼습니다!")
+        return true
+    default:
+        print("졌습니다!")
+        return true
```
게임의 종료조건이 이기거나 진 경우이기에 Bool 타입을 보내주는 것도 하나의 방법이지만

저 같은경우 어떤 상태인지 명확하게 전달하는 것을 선호해 게임 결과에 대한 상태를 보내줄 것 같습니다!

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
 //  Copyright © yagom academy. All rights reserved.
 // 
 
+//TODO: 함수이름
```
사용하지 않은 주석은 지워 주셔야 합니다 🤔

WhalesJin replied:
네!! 지웠습니다!! 알려주셔서 감사해요 릴라 😃

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
 
+//TODO: 함수이름
+
 import Foundation
```
Foundation 을 import 해주셨네요 Foundation 을 어떤 이유로 import 해주셨을 까요?

사소한 코드라도 사용을 했더라면 그 코드에 대한 이유가 있어야 합니다

agilestarskim replied:
Foundation의 역할은 대략적으로 다음과 같습니다.
* 텍스트 가공
* 날짜 및 시간 계산
* 정렬 및 필터링
* 네트워킹

현재 코드에서는 위와 같은 부가적인 기능이 필요없는 것을 확인하고 해당 구문을 삭제했습니다.

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
 import Foundation
 
-print("Hello, World!")
+gameLoop: while true {
```
`labeled statement` 를 사용해주셨군요👍🏻
저도 사용해본적이 없던 문법이었는데 배워갑니다 🙇🏻‍♂️

WhalesJin replied:
저희도 이번에 같이 공부하면서 찾았습니다!!
바로 써 먹을 수 있어서 너무 재밌었는데
함수들을 리팩토링 하고, 숫자를 이름들로 바꾸는 과정에서 지우게 되어서
많이 아쉬웠습니다 🥲

그렇지만 배웠으니 다음에 또 써먹어보겠습니다!!!

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+    let computer = String(Int.random(in: 1...3))
+    let input = readLine()
+    
+    switch input {
```
현재 gameLoop 라는 기능 안에서 

1. 메뉴 출력
2. 컴퓨터의 패 랜덤 설정
3. 사용자에게 입력 받기
4. 입력받은 값 분기처리

다양한 기능을 하고 있네요 🤔

어떻게 분리를 해볼 수 있을까요?

WhalesJin replied:
1. 메뉴 출력 받는 `printMenu()`함수
2. 컴퓨터 패 설정하고 가위바위보 모양으로 바꿔주는 `generateComputerHandShape()`함수
3. 사용자에게 입력받는 `getMenuFromUser()`함수
4. 입력 받은 후 다음 동작을 결정하는 `executeMenu(by menuNumber: Int?)`함수

이렇게 분리해보았습니다!!

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+    case "0":
+        print("게임 종료")
+        break gameLoop
+    case "1", "2", "3":
```
해당 코드를 다른사람의 코드라 가정하고 보았을 때 0, 1, 2, 3 이 어떤 것인지 알 수 있을까요?

agilestarskim replied:
switch 문에서 menu를 기준으로 case를 분리했고, 위에서 메뉴를 출력할 때 "가위(1), 바위(2), 보(3)" 문구가 출력되어 오히려 enum HandShape로 바꾸는 것보다 숫자 그대로를 받는 것이 더 적절하다고 생각했습니다.

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+        print("게임 종료")
+        break gameLoop
+    case "1", "2", "3":
+        if let user = input, judgeGame(user, with: computer) {
+            break gameLoop
```
```suggestion
        if let user = input, 
           judgeGame(user, with: computer) {
            break gameLoop
```

저같은 경우 가독성을 위해 하나의 라인에 하나의 로직이 들어오 돌고 개행을 하는 편입니다 😄 

---

현재 judgeGame 의 결과가 true 여야 gameLoop 가 종료되고 있네요.
bool 타입으로 게임결과가 어떤 상태인지 명확하게 구분을 할 수 있을까요?

agilestarskim replied:
리팩토링하여 하나의 조건으로 끝냈습니다 👍

저희도 그 부분이 명확하지 않다고 판단하여 게임의 상태를 나타내는 변수를 선언하고 그 변수를 토대로 게임을 끝낼지 재시작할지 여부를 판단하는 로직을 추가했습니다.

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+    }
+}
+
+func judgeGame(_ user: String, with computer: String) -> Bool {
```
Naming에 대한 질문사항이 있으셔서 조언드립니다.

Naming 의 경우 어떤행동에 초점을 맞추는지에 따라 사람마다 작성하는 Naming 이 천차만별이라 생각합니다.
Minsup과 Whales 은 해당 메서드에 어떤 행동에 초점을 맞추고 judgeGame 이란 Naming 을 사용했을까요?

---

혹시 [접근제어 (Access Control)](https://jusung.gitbook.io/the-swift-language-guide/language-guide/25-access-control) 에 대해 알고 계신가요?

현재 judgeGame 의 접근레벨은 무엇일까요?

agilestarskim replied:
위 코드에서 접근제어를 설정하지 않았고 그러면 기본값으로 Internal로 설정됩니다.

Internal : 기본 접근레벨로 아무 접근레벨을 선언하지 않으면 Internal로 간주됩니다. Internal레벨로 선언되면 해당 모듈 전체에서 사용 가능합니다. (모듈: 폴더단위)

현재 바뀐 코드에서는 private으로 되어있어 외부에서 접근할 수 없게 수정하였습니다.

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+
+func judgeGame(_ user: String, with computer: String) -> Bool {
+    switch (user, computer) {
+    case ("1", "1"), ("2", "2"), ("3", "3"):
```
언급했다시피 1, 2, 3 만 보고 해당 숫자가 어떤걸 의미하는지 알 수 있을까요?

매직넘버 / 리터럴에 대해 한번 알아보시겠어요?

WhalesJin replied:
매직 넘버(`magic number`)는 코드에서 직접 사용되는 숫자값으로 `매직`이라는 용어는 코드를 읽는 사람에게 숫자의 의미가 명확하지 않고 왜 이 값이 들어갔는지 의미를 알기 힘들다는 의미로 쓰였다고 합니다.

이러한 값들에게 이름을 부여하기 위해 상수(static final)로 선언하는 등의 방법을 사용하고, 이름을 통하여 의미와 역할을 확실히 전달할 수 있게 만들면 좋습니다.

그래서 저희는 1,2,3 대신 가위,바위,보로 코드를 보여주기 위해 열거형으로 `HandShape`라는 타입을 만들었습니다.

juun97 commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -4,7 +4,39 @@
+func judgeGame(_ user: String, with computer: String) -> Bool {
+    switch (user, computer) {
+    case ("1", "1"), ("2", "2"), ("3", "3"):
+        print("비겼습니다!")
+        return false
+    case ("1", "3"), ("3", "2"), ("2", "1"):
+        print("이겼습니다!")
+        return true
+    default:
+        print("졌습니다!")
+        return true
```
게임의 종료조건이 이기거나 진 경우이기에 Bool 타입을 보내주는 것도 하나의 방법이지만

저 같은경우 어떤 상태인지 명확하게 전달하는 것을 선호해 게임 결과에 대한 상태를 보내줄 것 같습니다!

WhalesJin replied:
조언주신 부분 참고하여 `GameResult`라는 열거형 타입을 만들어서
결과에 대한 상태를 반환하는 함수로 수정했습니다!!

Commit:
refactor: 피드백에 대한 전반적인 수정

main에서 파일 분리 후 Game 파일에서 Game 클래스 생성, gameLoop 내용들을 여러 함수로 기능 분리, HandShape, GameResult 열거형 타입 생성

WhalesJin commented:
여러가지 조언 감사드립니다 릴라(@juun97)!!!
주신 조언들 참고하여 코드 수정하고, 주신 물음들도 공부해서 답변 올려봅니다!!!🙇🏻🙇🏻‍♀️

1️⃣ 기능분리

> 하나의 메서드는 하나의 기능을 담당해야 한다고 많이들 말합니다.
그렇다면 왜 그래야 하는지 생각을 해본적이 있을까요?
> 왜 하나의 메서드가 여러기능을 담당하면 안될까요? 간단한 로직들이라면 하나의 메서드에서 다 처리 해버려도 별 문제가 없지 않을까요?

- 여러 기능을 담당하는 메서드들로 코드를 짜다보면 코드의 중복이 발생할 수 있다. 중복된 하나의 기능을 메서드로 꺼내오면 코드의 중복을 줄이고 같은 기능이 필요한 다른 메서드를 만들 때 재사용할 수 있다.
- 미래의 예상치 못한 부작용 발생을 줄일 수 있다.
    <details close>
    <summary>예시코드</summary>
    
    ```swift
    func greeting() {
        print(self.name + "안녕하세요")
        self.name = "minsup" //-> 부작용 발생
    }

    func changeName(name: String) {
        self.name = name 
    }

    changeName(name: "whales")
    greeting() // "whales 안녕하세요"
    greeting() // "minsup 안녕하세요"
    ```
    </details>

2️⃣ 매직넘버 / 리터럴

> 현재 Minsup과 Whales이 작업하지 않았다고 가정하고 작업한 코드를 보았을 때 어떤 코드인지 알 수 있나요? 부가적인 설명없이 코드가 이해가 될까요?
짧아진 코드가 자신만 알아볼 수 있는 코드라면 과연 좋은 코드라고 할 수 있을까요?🤔

미션의 요구사항 부분에 집중을 하다보니 놓친 부분 같습니다!! 알려주셔서 감사해요 릴라!! 주신 부분 참고해서 코드 수정했습니다!!

3️⃣ Branch

> 현재 main Branch 에서 작업을 진행해주셨네요! 혹시 branch 에 대해 알고계시나요?
하나의 branch 에서만 작업을 하게 된다면 어떤 문제가 생기게 될까요?
branch 를 나눠서 관리하면서 생기는 장점에 대해 알아볼까요

 - 브랜치를 여러 개로 나누면 기능 추가 브랜치, 버그 수정 브랜치등 독립적인 작업을 수행할 수 있습니다.
 - 또한 브랜치 별로 커밋이력을 나눠 저장할 수 있기 때문에 커밋이력 관리를 효과적으로 할 수 있습니다.
 - main 브랜치는 프로그램의 전체를 병합하는 역할을 담당하는 브랜치가 되어야 위에 말한 내용들을 수행할 수 있습니다.
